### DISCUSSION TITLE
New Static Scene for player and enemies

### DISCUSSION THREAD

**Discussion by prageeth79** [godotengine/godot-proposals#9703] [Engine Core]
### Describe the project you are working on

2D RPG Game

### Describe the problem or limitation you are having in your project

when moving from scene to scene every child scene in the world scene gets reset. even the player

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

if one can make these child scene retain their state after leaving the scene then one enters to the scene they will have their properties not chanced after leaving the scene

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

having an option to scene to be persistent across all so it will not change when changing the scene

### If this enhancement will not be used often, can it be worked around with a few lines of script?

you need to add a global script and  have to configure it, have to be handle separately from the scene

### Is there a reason why this should be core and not an add-on in the asset library?

it is a feature that will be used for most of 2D games and it will be very helpful and make the cording easiar

**Comment by Calinou** [comment]
Moving to discussion, as there is no concrete technical implementation provided.

**Comment by theraot** [comment]
This reminds me of some discussion on making a persistence system, something like this: The scene would have helper nodes that will take care of saving/loading the state of the scene, and spawning any additional nodes needed for that. You would have to configure these helper nodes to specify what parts of the scene would be persisted. Then you can call a method on them to take an snapshot and save it. And they will automatically load the last snaptshot on ready.

*We can also imagine this working with the multiplayer system (which already have the means to syncronize a scene across the network). But you could have local persistence without networking, and it might be an stepping stone for adding multiplayer to a geme.*

---

Anyway, I have seen people using PackedScene to store a scene before unloading it, so they can then instantiate the saved PacketScene when the player comes back. It is a "perfect" save... Kind of. It has issues:

- It saves irrelevant things, and things that will not change from gameplay to gameplay.
- It might cause problems if you try to load a save on a different version of the game (in some cases the loaded scene has scripts that conflict with the scripts in the game, in others the save file might reference things that no longer exist).
- It is susceptible to code injection (i.e. somebody can send/share/publish a poisoned save file, they do not need to trick players into downloading a compromised version of the game, so your players might have a false sense of security thinking it is just a save file).

But hey, easy to use.

*It seems to me that static scene idea would have some of these issues, since it would have to work without knowing what is worth saving... So it would save everything.*

---

My current solution is a combination of three things:

- A data store for settings, with custom binary serialization. Mostly used from game options and the main scene.
- A data store for records and game progress, with a section for temporary data which is not persisted. Many scripts read and write here directly (I make it easy for designers to add stuff there, such as score, items, quest progress, etc), so objects can read these values when they spawn and adjust accordingly.
- And a signal bus with a signal for saving and another for loading, so anything can subscribe to these. And if some part of the game needs to save something special they can do so in response (e.g. if the game allows the player to draw, we might save and load a texture).

---

Hopefully this is food for thought for anybody who wants to make a generic version of such systems.
